src/transformers/models/maskformer/modeling_maskformer.py

```diff
@@ -0,0 +1,2463 @@
+        checkpoint=_CHECKPOINT_FOR_DOC,
+        output_type=MaskFormerOutput,
+        config_class=_CONFIG_FOR_DOC,
+        modality="vision",
+    )
+    def forward(
+        self,
+        pixel_values: Tensor,
+        pixel_mask: Optional[Tensor] = None,
+        output_hidden_states: Option[bool] = False,
+        return_dict: Optional[bool] = True,
+    ) -> MaskFormerOutput:
+
+        if pixel_values is None:
+            raise ValueError("You have to specify pixel_values")
+        batch_size, _, height, width = pixel_values.shape
+
+        if pixel_mask is None:
+            pixel_mask = torch.ones((batch_size, height, width), device=pixel_values.device)
+        pixel_level_module_output: MaskFormerPixelLevelModuleOutput = self.pixel_level_module(
+            pixel_values, output_hidden_states
+        )
+        image_features: Tensor = pixel_level_module_output.encoder_last_hidden_state
+        pixel_embeddings: Tensor = pixel_level_module_output.decoder_last_hidden_state
+
+        transformer_module_output: DetrDecoderOutput = self.transformer_module(image_features, output_hidden_states)
+        queries: Tensor = transformer_module_output.last_hidden_state
+
+        if not return_dict:
+            return (image_features, pixel_embeddings, queries)
+
+        return MaskFormerOutput(
+            encoder_last_hidden_state=image_features,
+            pixel_decoder_last_hidden_state=pixel_embeddings,
+            transformer_decoder_last_hidden_state=queries,
+            encoder_hidden_states=pixel_level_module_output.encoder_hidden_states if output_hidden_states else (),
+            pixel_decoder_hidden_states=pixel_level_module_output.decoder_hidden_states
+            if output_hidden_states
+            else (),
+            transformer_decoder_hidden_states=transformer_module_output.hidden_states if output_hidden_states else (),
+        )
+
+
+class MaskFormerForInstanceSegmentation(MaskFormerPretrainedModel):
+    def __init__(self, config: MaskFormerConfig):
+        super().__init__(config)
+        self.model = MaskFormerModel(config)
+        hidden_size: int = config.detr_config.hidden_size
+        # + 1 because we add the "null" class
+        self.class_predictor = nn.Linear(hidden_size, config.num_labels + 1)
+        self.mask_embedder = MaskformerMLPPredictionHead(hidden_size, hidden_size, config.mask_feature_size)
+
+        self.matcher = MaskFormerHungarianMatcher(
+            cost_class=1.0, cost_dice=config.dice_weight, cost_mask=config.mask_weight
+        )
+
+        self.weight_dict: Dict[str, float] = {
+            "loss_cross_entropy": config.cross_entropy_weight,
+            "loss_mask": config.mask_weight,
+            "loss_dice": config.dice_weight,
+        }
+
+        self.criterion = MaskFormerLoss(
+            config.num_labels,
+            matcher=self.matcher,
+            weight_dict=self.weight_dict,
+            eos_coef=config.no_object_weight,
+        )
```

Comment:
(just a question) wondering if these objects can be encapsulated into a `MaskFormerSegmentationModule` (to match the figure in the paper)

Reply:
I'd say `MaskFormerForInstanceSegmentation` could be `MaskFormerSegmentationModule` but if we really want to match the figure in the paper we need to bring back the code from `post_process_segmentation` from the feature extractor